torch/package/package_exporter.py

```diff
@@ -588,6 +709,112 @@ def _write(self, filename, str_or_bytes):
             str_or_bytes = str_or_bytes.encode("utf-8")
         self.zip_file.write_record(filename, str_or_bytes, len(str_or_bytes))
 
+    def _validate_dependency_graph(self):
+        # 1. No modules should be denied.
+        # 2. No broken modules (we should have been able to retrieve source for everything interned).
+        # 3. All modules should have an associated action.
+        # 4. All patterns for which allow_empty=False have been matched at least once.
+        denied = set()
+        broken = {}
+        unhandled = set()
+        for module_name, attrs in self.dependency_graph.nodes.items():
+            if attrs.get("action") == _ModuleProviderAction.DENY:
+                denied.add(module_name)
+
+            if attrs.get("broken") is True:
+                filename = attrs.get("filename")
+                if filename is None:
+                    broken_reason = "Module does not have a __file__ attribute set."
+                elif filename.endswith(tuple(importlib.machinery.EXTENSION_SUFFIXES)):
+                    broken_reason = (
+                        "Module is an C extension module, which is not supported in packaging. "
+                        "Extern/mock it, or refactor your code to avoid the dependency."
+                    )
+                else:
+                    broken_reason = f"Source file {filename} not found."
+                broken[module_name] = broken_reason
+
+            if attrs.get("action") is None:
+                unhandled.add(module_name)
+
+        if len(denied) != 0 or len(broken) != 0 or len(unhandled) != 0:
+            # build up the filter set
+            interns = set()
+            for module_name, attrs in self.dependency_graph.nodes.items():
+                if attrs.get("action") == _ModuleProviderAction.INTERN:
+                    interns.add(module_name)
+
+            include_filter = interns.copy()
+            for intern in interns:
+                for dep in self.dependency_graph.successors(intern):
+                    include_filter.add(dep)
+
+            raise PackagingError(denied, broken, unhandled, include_filter)
+
+        for pattern, _, allow_empty in self.patterns:
+            if not allow_empty and pattern not in self.matched_patterns:
+                raise EmptyMatchError(
+                    f"Exporter did not match any modules to {pattern}, which was marked as allow_empty=False"
+                )
```

Comment:
One of the goals of this PR is to aggregate and present all errors at once, but these `EmptyMatchErrors` are not aggregated and will still be presented one at a time. Could we aggregate these as well? That is, could we raise one exception with *all* unmatched patterns?

Reply:
I'll file a follow-up issue for this (it's a little bit more involved to change)